### PR TITLE
[CIVIS-9806] ENH update civis-python to v2.4.0; add uv

### PR DESCRIPTION
**Pull request by jacksonlee-civis**
We've just released civis-python v2.4.0, so it's time for another round of updating the datascience-python image. Summary of changes from the updated changelog:

- Python version updated to v3.12.7
- Core dependencies updated to latest versions:
  * awscli 2.17.37 -> 2.19.5
  * boto3 1.34.127 -> 1.35.58
  * civis 2.3.0 -> 2.4.0
  * numpy 2.0.0 -> 2.1.3
  * pandas 2.2.2 -> 2.2.3
  * scikit-learn 1.5.0 -> 1.5.2
  * scipy 1.13.1 -> 1.14.1
- uv added to the image

---

- [x] (For Civis employees only) Reference to a relevant ticket in the pull request title
- [x] Changelog entry added to `CHANGELOG.md` at the repo's root level
- [x] Description of change in the pull request description
- [x] If applicable, unit tests have been added and/or updated
- [x] The CircleCI builds have all passed

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -37,23 +37,21 @@ COPY requirements-full.txt .
...
   VERSION_MAJOR=8 \
-  VERSION_MINOR=0 \
-  VERSION_MICRO=1
-
-# Install the AWSCLI for moving match targets in the QC workflow.
-# See https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html#cliv2-linux-install
-RUN curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip" && \
-    unzip awscliv2.zip && \
-    ./aws/install && \
-    rm -rf aws awscliv2.zip
```
Comment (jacksonlee-civis) [inline Dockerfile:56]: Now specifying awscli in requirements-core.txt instead, see below

File: generate-requirements-full.sh [hunk 2]
```diff
@@ -2,4 +2,4 @@
 # Run this script to update requirements-core.txt.
 # It uses Docker to ensure that the environment matches what will be used in the production image.
 set -e
-docker compose run --rm pip-tools /bin/sh -c "pip install --upgrade pip-tools && pip-compile --output-file=requirements-full.txt --pip-args='--prefer-binary' --strip-extras --upgrade requirements-core.txt"
+docker compose run --rm pip-tools /bin/sh -c "apt-get update -y && apt-get install -y git && pip install --upgrade pip-tools && pip-compile --output-file=requirements-full.txt --pip-args='--prefer-binary' --strip-extras --upgrade requirements-core.txt"
```
Comment (jacksonlee-civis) [inline generate-requirements-full.sh:5]: Need `git` for installing awscli from its github source

**Comment by jacksonlee-civis** [comment]
The CircleCI build has passed, but I'm switching this pull request to the "draft" status as I'm running tests internally on Civis Platform. Once this PR is ready for review, I'll tag you ~@Drew-Pappas~ (going to spread out the workload instead of tagging Drew all the time 😄)

**Review by margaretott — APPROVED** [review]
LGTM!

**Pull request merged**